addons/nuxeo-core-binarymanager-cloud/nuxeo-core-binarymanager-s3/src/main/java/org/nuxeo/ecm/core/bulk/S3SetContentLengthAction.java

```diff
@@ -0,0 +1,178 @@
+
+    public static final String DISABLE_THUMBNAIL_SERVICE = "disableThumbnailComputation";
+
+    public static final String DISABLE_AUTO_CHECKOUT = "DisableAutoCheckOut";
+
+    public static final String DISABLE_DOMAIN_LISTENER = "disableDomainListener";
+
+    public static final String DISABLE_HTMLSANITIZER_LISTENER = "disableHtmlSanitizerListener";
+
+    public static final String DISABLE_AUTO_INDEXING = "disableAutoIndexing";
+
+    private static final String FORCE_OPTION = "force";
+
+    @Override public Topology getTopology(Map<String, String> options) {
+        return Topology.builder()
+                       .addComputation(SetContentLengthComputation::new,
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class SetContentLengthComputation extends AbstractBulkComputation {
+
+        private static final Logger log = LogManager.getLogger(SetContentLengthComputation.class);
+
+        public SetContentLengthComputation() {
+            super(ACTION_NAME);
+        }
+
+        protected String currentProvider;
+
+        @Override protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+            DownloadBlobGuard.enable(); // Avoid downloading main content blob during the transaction
+            boolean force = properties.containsKey(FORCE_OPTION) && Boolean.parseBoolean(
+                    properties.get(FORCE_OPTION).toString());
+            for (DocumentModel doc : loadDocuments(session, ids)) {
+                updateContentLength(session, doc, force);
+            }
+        }
+
+        protected void updateContentLength(CoreSession session, DocumentModel doc, boolean force) {
+            BlobInfo blobInfo = getBlobInfo(doc);
+            if (blobInfo == null) {
+                log.debug("Skipping {}: no blob", doc);
+                return;
+            }
+            if (blobInfo.length != null && blobInfo.length > 0 && !force) {
+                log.debug("Skipping {}, keeping existing length: {}", doc, blobInfo.length);
+                return;
+            }
+            long length = getContentLength(doc, blobInfo);
+            if (length == blobInfo.length) {
+                log.debug("Skipping {}, length is correct {}", doc, length);
+                return;
+            }
+            disableListeners(doc);
+            try {
+                log.info("Updating {}, length from {} to {}", doc, blobInfo.length, length);
+                blobInfo.length = length;
+                Blob blob = new SimpleManagedBlob(currentProvider, blobInfo);
+                doc.setProperty("file", "content", blob);
+            } catch (PropertyException e) {
+                log.warn("Cannot set content/length: {} of document: {}", length, doc.getId(), e);
+            }
+            session.saveDocument(doc);
```

Comment:
Can `session#saveDocument` be moved inside the try/catch in order to not save the document in case of failure?